stan/math/opencl/rev/vari.hpp

```diff
@@ -0,0 +1,204 @@
+#include <stan/math/rev/core/vari.hpp>
+#include <stan/math/opencl/is_matrix_cl.hpp>
+#include <stan/math/opencl/kernel_generator.hpp>
+
+namespace stan {
+namespace math {
+/**
+ * The variable implementation for `matrix_cl`.
+ *
+ * This class is complete (not abstract) and may be used for
+ * constants.
+ *
+ * A variable implementation is constructed with a constant
+ * value. It also stores the adjoint for storing the partial
+ * derivative with respect to the root of the derivative tree.
+ *
+ */
+template <typename T>
+class vari_value<T, require_kernel_expression_lhs_t<T>>
+    : public vari_base, public chainable_alloc {
+ public:
+  /**
+   * The adjoint of this variable, which is the partial derivative
+   * of this variable with respect to the root variable.
+   */
+  T adj_;
+
+  /**
+   * The value of this variable.
+   */
+  T val_;
+
+  /**
+   * Construct a matrix_cl variable implementation from a value. The
+   * adjoint is initialized to zero.
+   *
+   * All constructed variables are added to the stack. Variables
+   * should be constructed before variables on which they depend
+   * to insure proper partial derivative propagation.  During
+   * derivative propagation, the chain() method of each variable
+   * will be called in the reverse order of construction.
+   *
+   * @tparam S A `matrix_cl` or kernel generator expression type that is
+   * convertible to `value_type`
+   * @param x Value of the constructed variable.
+   */
+  template <typename S, require_convertible_t<S&, T>* = nullptr>
+  explicit vari_value(S&& x)
+      : chainable_alloc(),
+        adj_(constant(0, x.rows(), x.cols())),
+        val_(std::forward<S>(x)) {
+    ChainableStack::instance_->var_stack_.push_back(this);
+  }
+
+  /**
+   * Construct a matrix_cl variable implementation from an Eigen value. The
+   * adjoint is initialized to zero.
+   *
+   * All constructed variables are added to the stack. Variables
+   * should be constructed before variables on which they depend
+   * to insure proper partial derivative propagation.  During
+   * derivative propagation, the chain() method of each variable
+   * will be called in the reverse order of construction.
+   *
+   * @tparam S A dense Eigen value or expression that has same scalar type as T
+   * @param x Value of the constructed variable.
+   */
+  template <typename S, require_eigen_t<S>* = nullptr,
+            require_vt_same<T, S>* = nullptr>
+  explicit vari_value(const S& x)
+      : chainable_alloc(), adj_(constant(0, x.rows(), x.cols())), val_(x) {
+    ChainableStack::instance_->var_stack_.push_back(this);
+  }
+
+  /**
+   * Construct an matrix_cl variable implementation from a value. The
+   *  adjoint is initialized to zero and if `stacked` is `false` this vari
+   *  will be not be put on the var_stack. Instead it will only be put on
+   *  a stack to keep track of whether the adjoint needs to be set to zero.
+   *
+   * All constructed variables are added to a stack.  Variables
```

Comment:
This sentence isn't true if stacked == false.

Reply:
I think the specific wording `a stack` here makes this sentence true. Things are either added to the `var_stack_` or `var_nochain_stack_`